Async

kc7851 opened the pull request:
아직 갈길이 머네요... ㅜㅜ

kc7851 committed:
App test 작성

kc7851 committed:
App test 구현

kc7851 committed:
App tset 작성

kc7851 committed:
App test 구현

kc7851 committed:
Regions test 작성

kc7851 committed:
Regions test 구현

kc7851 committed:
Categories test 작성

kc7851 committed:
Categories test 구현

kc7851 committed:
App 리팩토링

Categories component를 사용합니다.

kc7851 committed:
Restaurants test 작성

kc7851 committed:
Restaurants test 구현

kc7851 committed:
CategoriesContainer test 작성

kc7851 committed:
CategoriesContainer test 구현

kc7851 committed:
Categories test 작성

props로 categories를 넘겨줍니다.

kc7851 committed:
Categories test 구현

CategoriesContainer에서 props로 categories 배열을 넘겨줍니다.

kc7851 committed:
App 리팩토링

Categories에서 CategoriesContainer component로 변경하였습니다.

kc7851 committed:
Categories test 작성

kc7851 committed:
Categories test 구현

kc7851 committed:
Categories test 리팩토링

kc7851 committed:
Categories test 작성

kc7851 committed:
Categories test 구현

kc7851 committed:
App, CategoriesContainer test 리팩토링

정확한 문자열에서 정규식으로 변경

kc7851 committed:
redux store 생성

kc7851 committed:
recuder test 작성

updateSeleted test

kc7851 committed:
reducer test 구현

kc7851 committed:
CategoriesContainer test 작성

kc7851 committed:
CategoriesContainer test 구현

kc7851 committed:
CategoriesContainer test 작성

kc7851 committed:
CategoriesContainer test 구현

kc7851 committed:
App test 리팩토링

useDispatch, useSelector를 mocking하도록 하였습니다.

kc7851 committed:
Index 리팩토링

App 아래에서 redux를 사용하기위해 Provider로 감싸줍니다.

kc7851 committed:
reducer test 리팩토링

action creator를 사용하도록 변경하였습니다.

kc7851 committed:
actions 작성

kc7851 committed:
CategoriesContainer 리팩토링

action을 사용하도록 변경하였습니다.

kc7851 committed:
Categories 리팩토링

onClick 함수에 id인자를 넘겨주도록 변경했습니다.

kc7851 committed:
Categories test case 추가

kc7851 committed:
CategoriesContainer test 변경

selector에서 selected categoryId값을 주어서 표시가 나타나도록 하였습니다.

kc7851 committed:
reduce initialState 추가

kc7851 committed:
RegionsContainer test 작성

kc7851 committed:
RegionsContainer test 구현

kc7851 committed:
Regions 리팩토링

RegionsContainer에서 넘겨준 props를 이용하도록 변경하였습니다.

kc7851 committed:
App 리팩토링

Regions를 RegionsContainer로 변경하였습니다.

kc7851 committed:
Regions test case 추가

kc7851 committed:
reducer test 작성

kc7851 committed:
recuder test 구현

kc7851 committed:
action updateSelected 변경

1. 하나의 action으로 두개의 변경사항을 처리하기위해 updateSelected변경하였습니다.

2. Categories, Regions에서 함수호출 인자를 변경하였습니다.

3. CategoriesContainer, RegionsContainer에서 action에 넘겨주는 인자를 변경하였습니다.

kc7851 committed:
RegionsContainer test 오류 수정

kc7851 committed:
CategoriesContainer test 작성

kc7851 committed:
Categories test 구현

Container에서 Categoriees에 넘겨주는 categories를 useSelector로 store에서 받도록 변경했습니다.

kc7851 committed:
ResionsContainer test 작성

kc7851 committed:
RegionsContainer test 구현

Container에서 Rigions에 넘겨주는 regions를 useSelector로 store에서 받도록 변경했습니다.

kc7851 committed:
App test 오류 수정

기본 값을 변경하였습니다.

kc7851 committed:
reducer test case 추가

선언되지 않은 action이 들어왔을 때와, state가 undefined일 경우의 테스트를 추가하였습니다.

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,61 @@
+
+  useSelector.mockImplementation((selector) => selector({
+    selected: {},
+    categories: [
```
테스트용 데이터를 fixture라고 부르는데 따로 파일로 분리하면 좋을 것 같아요

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,61 @@
+  useDispatch.mockImplementation(() => dispatch);
+
+  useSelector.mockImplementation((selector) => selector({
+    selected: {},
```
selected는 true 혹은 false만 담고 있을 것 같아요. 선택된 카테고리 혹은 선택된 지역을 표시하고 싶으셨던 거죠

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,61 @@
+      <App />,
+    );
+
+    expect(getByText(/서울/)).not.toBeNull();
```
regions를 이용해서 표현하면 어떨까요?

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,61 @@
+      <App />,
+    );
+
+    expect(getByText(/한식/)).not.toBeNull();
```
여기도 categories를 이용해서 표현해주는건 어떨까요?

hannut91 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,64 @@
+jest.mock('react-redux');
+
+describe('CategoriesContainer', () => {
+  const dispatch = jest.fn();
```
dispatch가 테스트 여러곳에서 쓰일경우 초기화하지 않으면 문제가 발생할 수 있어요 

```js
beforeEach(() => {
  dispatch.mockClear
});
```

hannut91 reviewed (changes requested):
좋네요 ㅎㅎㅎ TDD하실 때 리팩터링을 빠트리지 마시고 꼭하세요 ㅎㅎ

kc7851 committed:
fixtures로 분리

테스트에 자주 사용되는 변수를 fixtures로 분리하여 재사용이 쉽도록 만듭니다.

kc7851 committed:
App test 리팩토링

1. fixture를 사용하도록 변경했습니다.

2. 배열의 forEach를 이용해 테스트하도록 변경했습니다.

kc7851 committed:
Categories test 리팩토링

categories를 fixture로 변경했습니다.

kc7851 committed:
CategoriesContainer test 리팩토링

1. fixture를 사용하도록 변경하였습니다.

2. beforeEach로 dispatch를 초기화하도록 하였습니다.

kc7851 committed:
Regions test 리팩토링

1. fixture를 사용하도록 변경하였습니다.

2. 배열을 forEach를 사용하도록 변경하였습니다.

kc7851 committed:
RegionsContainer test 리팩토링

1. fixture를 사용하도록 변경하였습니다.

2. 배열을 forEach를 사용하도록 변경하였습니다.

kc7851 committed:
RestaurantsContainer test 작성

kc7851 committed:
RestaurantsContainer test 작성

kc7851 committed:
restauratns fixture 생성

kc7851 committed:
Restaurants test 작성

kc7851 committed:
Restaurants test 구현

kc7851 committed:
App test case 추가

kc7851 committed:
App 리팩토링

RestaurantsContainer를 사용하도록 변경하였습니다.

kc7851 committed:
RestaurantsContainer test 오타수정

restuarants -> restaurants로 수정하였습니다.

kc7851 committed:
reducer initialState 수정

restaurants를 추가하였습니다.

kc7851 committed:
App 리팩토링

initialState에 있는 categories를 없애고 외부의 값으로 categories를 업데이트하도록 변경합니다.

kc7851 committed:
App 리팩토링

action creator를 사용하도록 변경하였습니다.

kc7851 committed:
action 추가

setCategories action을 추가합니다.

kc7851 committed:
reducer test 작성

kc7851 committed:
reducer test 구현

kc7851 committed:
App 리팩토링

Regions도 Categories와 마찬가지로 외부에서 값을 가져와 설정하도록 변경합니다.

kc7851 committed:
action 추가

setRegions action creator를 추가하였습니다.

kc7851 committed:
reducer initialState 변경

regions의 초기값을 빈배열로 변경하였습니다.

kc7851 committed:
recuder test 작성

kc7851 committed:
reducer test 구현

kc7851 committed:
actions 오타 수정

kc7851 committed:
App 기능추가

 setRestaurants action을 추가 하였습니다.

kc7851 committed:
action 추가

setRestaurants action을 만들었습니다.

kc7851 committed:
reducer test 작성

kc7851 committed:
reducer test 구현

kc7851 committed:
recuder 리팩토링

initialState restaurants을 빈 배열을 주도록 변경하였습니다.

kc7851 committed:
redux-thunk 적용

store에 middleware로 redux-thunk를 사용하도록 하였습니다.

kc7851 committed:
App 리팩토링

api로 regions를 받아서 적용할 수 있도록 action을 변경하였습니다.

kc7851 committed:
action 구현

loadRegions action을 작성하였습니다.

kc7851 committed:
sevices/api.js 생성

fetch를 하기위한 api를 작성하였습니다.

kc7851 committed:
actions 수정

api의 fetchRegions를 import 하도록 하였습니다.

kc7851 committed:
App 리팩토링

api로 받아온 categories를 적용하도록 변경하였습니다.

kc7851 committed:
action 구현

loadCategories action을 구현하였습니다.

kc7851 committed:
api 작성

fetchCategories를 작성하였습니다.

kc7851 committed:
App test 리팩토링

api를 mocking하도록 하였습니다.

kc7851 committed:
오타 수정

sevices를 services로 수정하였습니다.

kc7851 committed:
App 리팩토링

setRestaurants에서 loadRestaurants 액션을 사용하도록 변경하였습니다.

kc7851 committed:
action 추가

loadRestaurants action을 구현하였습니다.

kc7851 committed:
api 작성

fetchRestaurants를 구현하였습니다.

kc7851 committed:
api mock 작성

fetchRestaurants에 대한 mock 함수를 만들었습니다.

kc7851 committed:
redux-mock-store 설치

비동기액션을 테스트하기 위해 설치하였습니다.

kc7851 committed:
App test case 추가

비동기액션 테스트를 추가하였습니다.

kc7851 committed:
action 수정

fetchRestaurants에서 categoryId가 falsy일 때만 동작하지 않도록 수정하였습니다.

kc7851 committed:
App test 수정

loadRestaurants에서 region값이 없을 때 case를 삭제하였습니다.

kc7851 committed:
api test case 추가

jest.fn().mockResolvedValue를 여러번 쓰다보니 await를 안해주면 가장 마지막에 실행하는 RESTAURANTS 값이 return되어서 모두 await를 해주었습니다.

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,138 @@
+    });
+  });
+
+  describe('loadCategories', () => {
+    const store = mockStore({});
+
+    it('categories를 fetch 한다.', async () => {
+      await store.dispatch(loadCategories());
+      const actions = await store.getActions();
+
+      expect(actions[0].type).toBe('setCategories');
+    });
+  });
```
App 컴포넌트에서 액션 테스트하는 것은 따로 분리하는게 좋을 것 같아요.

kc7851 replied:
actions.test.js 로 분리하면 되겠죠??

hannut91 replied:
넵

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+export default function Categories({ selected, categories, onClick }) {
```
`selected`는 true, false만 가질 것 처럼 보여요. selectedCategoryId는 어떨까요?

kc7851 replied:
selectedCategoryId 이게 좀 더 명확해 보이네요

hannut91 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,66 @@
+  });
+
+  context('without selected id', () => {
+    it('(V)가 표시되지 않는다..', () => {
```
`..` 점이 2개 들어가 있네요

hannut91 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,62 @@
+        <CategoriesContainer />,
+      );
+
+      expect(getByText(/한식/)).not.toBeNull();
+      expect(getByText(/중식/)).not.toBeNull();
+      expect(getByText(/일식/)).not.toBeNull();
+      expect(getByText(/양식/)).not.toBeNull();
+      expect(getByText(/분식/)).not.toBeNull();
```
여기도 `fixture`를 사용하면 간단하게 표현이 가능하겠죠?

hannut91 commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,58 @@
+        <RegionsContainer />,
+      );
+
+      regions.forEach((region) => {
+        const { name } = region;
+        const regExp = new RegExp(name);
+        expect(getByText(regExp)).not.toBeNull();
+      });
```
`regExp`라는 변수이름이 코드를 이해하는데 도움이 되지 않는 것 같아요. 우변이 이미 충분히 설명하고 있죠. 그리고 이름만 사용한다면 디스트럭처링을 바로 해도 괜찮을 것 같아요.

```js
regions.forEach(({ name }) => {
  expect(getByText(RegExp(name)).toBeInTheDocument();
});
```

See also
  - https://github.com/hannut91/refactoring-practice/tree/master/6.7-rename-variable

kc7851 replied:
오 아주 좋네요!

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,52 @@
+  restaurants: [],
+};
+
+export default function reducer(state = initialState, action) {
```
`if`나 `switch` 없이 이 문제를 해결할 수 있을까요?

kc7851 replied:
객체로 만들어서 분리하네요. ㅎㅎ 이거.. 어디서 본거같아서 예전에 10주 프로젝트할때 했던 리액트 네이티브코드 찾아보니 그렇게 했었네요. 아샬님 frontEnd TDD repo에 있는거 보고서. 그리고 redex-thunk도 사용하고 비동기 액션도 사용했던데, 아예 기억이 안났었네요 ㅜㅜ

kc7851 committed:
리뷰 수정사항 반영

1. 불필요한 변수 제거. 디스트럭쳐링 활용.

2. 변수명을 좀 더 의도가 드러나도록 변경.

kc7851 committed:
App test 리팩토링

비동기함수와 관련한 테스트를 actions test에서 하도록 변경하였습니다.

kc7851 committed:
reducer 리팩토링

풀이강좌 참조해서 리팩토링 하였습니다.

kc7851 committed:
utils.js 생성

get, equal 함수와 test를 만들었습니다.

kc7851 committed:
useSelector 리팩토링

useSelector에서 utils에서 만든 get을 이용하도록 변경하였습니다.

kc7851 commented:
utils에 get이나 equal에 object를 넘겨주지 않았는데, return에서 어떻게 obj를 사용할 수 있나요?

useSelector((state)) => state.regions)에서
useSelector(get('regions')) 이렇게 바뀐건데... 함수의 인자로 넘어오는 객체가 자동으로 들어가는건가요?

hannut91 commented on src/actions.test.js:
```diff
@@ -0,0 +1,71 @@
+  describe('loadRegions', () => {
+    const store = mockStore({});
+
+    it('regions를 fetch 한다.', async () => {
```
regions룰 fetch하는지에 대해서 테스트에서 하나도 드러나지 않는 것 같아요. setRegions액션이 실행된다로 이름을 변경하던지 아니면 실제로 fetch하는지 테스트를 변경하던지 하면 좋을 것 같아요

hannut91 commented on src/actions.test.js:
```diff
@@ -0,0 +1,71 @@
+  });
+
+  describe('loadRestaurants', () => {
+    const withoutCategoryIdStore = mockStore({
```
카테고리 아이디 없이 실행한다는 것은 우변 코드에서 이메 충분히 설명하고 있는 것 같아요. 간단하게 store라고만 해줘도 될 것 같아요.

hannut91 replied:
그리고 여기 위치가 이상하네요 컨텍스트 안에 들어가야 할 것 같아요

hannut91 commented on src/actions.test.js:
```diff
@@ -0,0 +1,71 @@
+      });
+    });
+
+    const withSelectedStore = mockStore({
```
얘도 컨텍스트안에 들어가야 할 것 같아요

hannut91 reviewed (approved):
고생하셨습니다! TDD를 하실 때 리팩터링을 꼭 같이 하면 좋을 것 같습니다. 구현에 대한 것은 막힘없이 잘 하셔서 코드의 표현력만 더 신경쓰시면 좋을 것 같아요

hannut91 commented:
이번 과제를 진행하면서 무엇을 배우셨나요? 다음 과제에서 더 신경써서 하고 싶으신게 있나요?

kc7851 committed:
리뷰 반영

1. test 이름 변경. test가 내용을 잘 표현할 수 있도록 해야 함.

2. 변수 위치 변경

kc7851 commented:
# 5주차 비동기 회고

## 배운 점
1. Promise와 async-await의 사용법을 배웠습니다. 
2. 비동기 액션을 사용하는 방법을 배웠습니다.
3. api를 이용하여 data를 가져오는 방법을 배웠습니다.
4. 리뷰를 통해 fixtures 사용을 배웠습니다.
5. mocking하여 test하는 방법을 배웠습니다. (api, react-redux, thunk)

## 느낀 점
1. 처음 생각한것보다 훨씬 더 많은 시간이 들었습니다. 역시 첫 시작이 가장 어렵습니다.
2. mocking하여 test하는 방법을 몰라서 어려운 부분이 많았습니다.
3. api를 직접 사용해서 코드를 작성하니 웹 애플리케이션 느낌이 나서 좋았습니다.
4. 최대한 복붙을 안하려고 하는편인데, 1달 동안 조금은 익숙해진 느낌이 들어서 좋았습니다.

## 아쉬운 점
1. tdd 사이클에서 리팩토링을 계속 건너띈 점. 
2. 클로저 함수에 대한 이해가 낮은 점.
3. 백지에서 코드를 작성할 때 어떤걸 import 해야하는지 막히는 부분이 있습니다.

## 앞으로
1. 틈 날때 마다 뒤 돌아보면서 코드를 점검하는 습관을 길러야 겠습니다.
2. 클로저 함수에 대해 공부를 해야겠습니다.
3. 익숙하지 않은 라이브러리를 익숙하게 만들어야 겠습니다.

kc7851 commented:
어랏..ㅋㅋ 이거 마크다운 지원하는군요 ㅋㅋ 모르고 그냥 작성했는데...

hannut91 commented:
앞으로가 기대되는군요! 클로저 함수는 어어어어엄청 많이 쓰이므로 공부하시면 좋습니다 ㅎㅎ
